src/preprocessing/passes/synth_rew_rules.cpp

```diff
@@ -18,143 +18,436 @@
 #include "options/base_options.h"
 #include "options/quantifiers_options.h"
 #include "printer/printer.h"
+#include "printer/sygus_print_callback.h"
 #include "theory/quantifiers/candidate_rewrite_database.h"
+#include "theory/quantifiers/quantifiers_attributes.h"
+#include "theory/quantifiers/sygus/sygus_grammar_cons.h"
+#include "theory/quantifiers/term_canonize.h"
+#include "theory/quantifiers/term_util.h"
 
 using namespace std;
+using namespace CVC4::kind;
 
 namespace CVC4 {
 namespace preprocessing {
 namespace passes {
 
-// Attribute for whether we have computed rewrite rules for a given term.
-// Notice that this currently must be a global attribute, since if
-// we've computed rewrites for a term, we should not compute rewrites for the
-// same term in a subcall to another SmtEngine (for instance, when using
-// "exact" equivalence checking).
-struct SynthRrComputedAttributeId
-{
-};
-typedef expr::Attribute<SynthRrComputedAttributeId, bool>
-    SynthRrComputedAttribute;
-
 SynthRewRulesPass::SynthRewRulesPass(PreprocessingPassContext* preprocContext)
     : PreprocessingPass(preprocContext, "synth-rr"){};
 
 PreprocessingPassResult SynthRewRulesPass::applyInternal(
     AssertionPipeline* assertionsToPreprocess)
 {
-  Trace("synth-rr-pass") << "Synthesize rewrite rules from assertions..."
+  Trace("srs-input") << "Synthesize rewrite rules from assertions..."
                          << std::endl;
   std::vector<Node>& assertions = assertionsToPreprocess->ref();
+  if (assertions.empty())
+  {
+    return PreprocessingPassResult::NO_CONFLICT;
+  }
 
-  // compute the variables we will be sampling
-  std::vector<Node> vars;
-  unsigned nsamples = options::sygusSamples();
-
-  Options& nodeManagerOptions = NodeManager::currentNM()->getOptions();
-
-  // attribute to mark processed terms
-  SynthRrComputedAttribute srrca;
+  NodeManager* nm = NodeManager::currentNM();
 
   // initialize the candidate rewrite
-  std::unique_ptr<theory::quantifiers::CandidateRewriteDatabaseGen> crdg;
   std::unordered_map<TNode, bool, TNodeHashFunction> visited;
   std::unordered_map<TNode, bool, TNodeHashFunction>::iterator it;
   std::vector<TNode> visit;
-  // two passes: the first collects the variables, the second registers the
-  // terms
-  for (unsigned r = 0; r < 2; r++)
+  // Get all usable terms from the input. A term is usable if it does not
+  // contain a quantified subterm
+  std::vector<Node> terms;
+  // all variables (free constants) appearing in the input
+  std::vector<Node> vars;
+
+  // We will generate a fixed number of variables per type. These are the
+  // variables that appear as free variables in the rewrites we generate.
+  unsigned nvars = options::sygusRewSynthInputNVars();
+  // must have at least one variable per type (e.g. true, false for Bool)
```

Comment:
say this also in the header

Reply:
This has been addressed, see line 48 synth_rew_rules.h